src/analysis_and_optimization/Partial_evaluator.ml

```diff
@@ -677,9 +712,38 @@ let rec eval_expr (e : Middle.expr_typed_located) =
             Lit (Int, Int.to_string (Bool.to_int (r1 <> 0. || r2 <> 0.)))
         | e1', e2' -> EOr (e1', e2') )
       | Indexed (e, l) ->
-          (* TODO: do something clever with array and matrix expressions here?
-  Note  that we could also constant fold array sizes if we keep those around on declarations. *)
-          Indexed (eval_expr e, List.map ~f:eval_idx l) ) }
+          let rec eval_indexed_expr expr =
+            let rec eval_indexed_expr_one_step expr =
+              match expr with
+              | Indexed (e, l) -> (
+                match l with
+                | [] -> e.expr
+                | [_] -> (
+                  match e.expr with
+                  | FunApp (t, f, elts)
+                    when f = string_of_internal_fn FnMakeArray
+                         || f = string_of_internal_fn FnMakeRowVec -> (
+                    match get_indices (List.length elts) (List.hd_exn l) with
+                    | None -> expr
+                    | Some indices ->
+                        if List.length indices = 1 then
```

Comment:
I think this might look nicer if included in the pattern matching:
```ocaml
| Some [index] -> (List.nth_exn elts index).expr
| Some indices -> FunApp (t, f, List.map indices ~f:(List.nth_exn elts))
```